tsdate/evaluation.py

```diff
@@ -1,254 +1,354 @@
+        for i, j in product(n_i, n_j):
+            query_node.append(i)
+            target_node.append(j)
+            shared_span.append(span)
+
+    numer = scipy.sparse.coo_matrix(
+        (shared_span, (query_node, target_node)),
+        shape=(ts.num_nodes, other.num_nodes),
+    ).tocsr()
+
+    return numer
+
+
+def match_node_ages(ts, other):
+    """
+    For each node in `ts`, return the age of a matched node from `other`.  Node
+    matching is accomplished by calculating the intervals over which pairs of
+    nodes (one from `ts`, one from `other`) subtend the same set of samples.
+
+    Returns three vectors of length `ts.num_nodes`: the age of the best
+    matching node in `other` (e.g.  with the longest shared span); the
+    proportion of the node span in `ts` that is covered by the best match; and
+    the id of the best match in `other`.
+
+    If either tree sequence contains unary nodes, then there may be multiple
+    matches with the same span for a single node. In this case, the returned
+    match is the node with the smallest integer id.
+    """
+
+    shared_spans = shared_node_spans(ts, other)
+    matched_span = shared_spans.max(axis=1).todense().A1
+    best_match = shared_spans.argmax(axis=1).A1
+    # NB: if there are multiple nodes with the largest span in a row,
+    # argmax returns the node with the smallest integer id
+    matched_time = other.nodes_time[best_match]
+
+    best_match[matched_span == 0] = tskit.NULL
+    matched_time[matched_span == 0] = np.nan
+
+    return matched_time, matched_span, best_match
+
+
+def node_spans(ts):
+    """
+    Returns the array of "node spans", i.e., the `j`th entry gives
+    the total span over which node `j` is in the tree (i.e., does
+    not have 'missing data' there).
+    """
+    child_spans = np.bincount(
+        ts.edges_child,
+        weights=ts.edges_right - ts.edges_left,
+        minlength=ts.num_nodes,
+    )
+    for t in ts.trees():
+        span = t.span
+        for r in t.roots:
+            # do this check to exempt 'missing data'
+            if t.num_children(r) > 0:
+                child_spans[r] += span
+    return child_spans
+
+
+def tree_discrepancy(ts, other):
+    """
+    For two tree sequences `ts` and `other`,
+    this method returns two values, as a tuple:
+    1. the fraction of the total span of `ts` over which each nodes' descendant
+    sample set does not match its' best match's descendant sample set.
+    2. The root mean squared difference
+    between the times of the nodes in `ts`
+    and times of their best matching nodes in `other`,
+    with the average weighted by the nodes' spans in `ts`.
+
+    This is done as follows:
+
+    For each node in `ts` the best matching node(s) from `other`
+    has the longest matching span using `shared_node_spans`.
+    If there are multiple matches with the same longest shared span
+    for a single node, the best match is the match that is closest in time.

```

Comment:
This agrees with the two-step procedure I wrote in earlier comment, but I don't think that's exactly what's happening below?